### DISCUSSION TITLE
How to achieve SSG?

### DISCUSSION THREAD

**Discussion by ghost** [preactjs/preact#4228] [Q&A] [answered]
Hello, I want to create static pages with Preact and have began to [toy](https://github.com/RickStanley/poc-ssg-preact) around with [Vike](https://github.com/vikejs/vike).

The thing is, Vike has its own complexity and things that I can not work around at this moment. Then I thought I'd go look for a scaffold with: Vite + Preact SSG project, but couldn't find one. Ideally I did not want to make one myself. The `npm init preact` one uses Vike for a SSG setup.

I have some specific needs for my configuration (see: https://github.com/vikejs/vike/discussions/1345, if interested), and I would rather begin with a minimal setting to see how things work out.

As I've said, I looked for examples, and found some for SSR, for instance: https://github.com/bluwy/create-vite-extra/tree/master/template-ssr-preact

And others for SSG, but not specifically for Preact, example: https://github.com/vitejs/vite-plugin-vue/tree/main/playground/ssr-vue

And finally, today I found out about [preact-iso](https://github.com/preactjs/preact-iso), but I'm a bit unsure how to approach the solution, I feel like this is an already solved problem, and I have the tools to assemble it, I just can't quite figure it out; I'd appreciate any guidance.

**Comment by JoviDeCroock** [comment]
We have the SSR starter in `create-preact` https://github.com/preactjs/create-preact/tree/master/templates/ssr this uses `vike` so it can be used for SSG as well.

Apart from that SSG could be achieved by invoking `preact-render-to-string` at build time

**Comment by rschristian** [comment, marked as answer]
Not released yet, but you might want to take a look at https://github.com/preactjs/preset-vite/pull/86 (and if you do, let me know how it goes!)

Edit: This has since been released. You can use it by creating a new project with `create-preact` or by setting it up yourself with `@preact/preset-vite`.

**Reply by ghost** [reply]
Hi @rschristian, thanks for sharing this.

I tried it locally but I couldn't get it to work.

No repository at this moment, but I'll put my steps here:
1. clone https://github.com/preactjs/preset-vite locally and link it with `npm link`;
2. go to a Vite + Preact project and run `npm link @preact/preset-vite` (it copied the contents "[demo](https://github.com/preactjs/preset-vite/tree/dce372da44942391a056c5cf270cb5d9a09f88df/demo)" folder over, to make it easier);
3. create a script in package.json to build using `vite build`;
4. found a error saying that "rs-module-lexer is not a module (ESM) package", a `npm i -D rs-module-lexer` fixed it;
5. ran build;
6. nothing broke, but nothing was statically rendered either.

Sorry if it doesn't tell much, but I wanted to share my attempt.

**Reply by rschristian** [reply]
> found a error saying that "rs-module-lexer is not a module (ESM) package", a `npm i -D rs-module-lexer` fixed it;

Hm, that shouldn't fix it to my knowledge? I did catch this last night and (locally) fixed it, needs a tiny adjustment:

```diff
-import { parseAsync as moduleLexerParseAsync } from "rs-module-lexer";
+import rs from 'rs-module-lexer';
...
-		const { output } = await moduleLexerParseAsync({
+		const { output } = await rs.parseAsync({
```

Sorry about that.

> clone https://github.com/preactjs/preset-vite locally and link it with `npm link`;

Generally I'd recommend against `npm link` (and equivalents), they're really poor for testing packages locally. I can fully recommend [yalc](https://npm.im/yalc) though, it's perfect for this and fixes the issues of using local symlinks (for example, `npm link` will work if a package is listed as a `devDependency` when it really needs to be a `dependency` -- `yalc` fixes this).

> go to a Vite + Preact project and run npm link @preact/preset-vite (it copied the contents "[demo](https://github.com/preactjs/preset-vite/tree/dce372da44942391a056c5cf270cb5d9a09f88df/demo)" folder over, to make it easier);

Did you copy the demo folder over from the PR? Just wondering as demo folder from that PR should do _something_, either error out or succeed. Prerendering isn't enabled by default though, so it would do nothing on a project without it enabled.

I appreciate you taking a look and taking the time to reply with your results!

**Reply by ghost** [reply]
> Did you copy the demo folder over from the PR? Just wondering as demo folder from that PR should do something, either error out or succeed. Prerendering isn't enabled by default though, so it would do nothing on a project without it enabled.

I copied the contents onto mine, respecting each path (of "demo") and its contents, manually.

> Prerendering isn't enabled by default though, so it would do nothing on a project without it enabled.

I figured it wasn't, from the "vite.config.js" demo (i. e. `{ prerender: { enabled: true } }`, so I did the same in my configuration.

I'll try it again, with `yalc` and use the "demo" content itself without modifying it.

> I appreciate you taking a look and taking the time to reply with your results!

No problem. :)

I'm tasked to find alternatives that satisfies our requirements, I have a big window to test each technology of my own choosing out. The last time I used Preact was 4-5 years ago, it was nice to see still going strong!

**Reply by rschristian** [reply]
> I copied the contents onto mine, respecting each path (of "demo") and its contents, manually.

Hm, odd! Was working for me last night in a newly create project with `create-preact` (with the same adjustments made that can be seen in that demo folder).

> I figured it wasn't, from the "vite.config.js" demo (i. e. { prerender: { enabled: true } }, so I did the same in my configuration.

Yeah that should be good to go then.

The output HTML (`index.html`, and any other routes it could discover) were still empty? Bizarre.

**Reply by ghost** [reply]
So, I tried it again. Steps:
1. clone https://github.com/preactjs/preset-vite locally (`git clone url_here -b feat/prerender-plugin`); 
2. change `'rs-module-lexer'` to be `import rs from 'rs-module-lexer';` and modifiy `moduleLexerParseAsync` accordingly;
3. execute `yalc publish`;
4. execute `cp -r preset-vite/demo my_demo`;
5. got to "my_demo";
6. execute `npm init -y`;
7. change project type to "module";
8. add dependencies and devDependencies: `preact`, `preact-iso`, `preact-render-to-string`, `typescript`, `vite`, `@preact/preset-vite`;
9. execute `yalc add @preact/preset-vite`;
10. remove "noEmit" from tsconfig;
11. change preact plugin import in "vite.config.ts" to `import preact from "@preact/preset-vite"`;
12. add script `"build": "vite build"`;
13. execute `npm run build`;
14. see crash.

Crash stack:

```bash
arch@archlinux ~/e/my_demo> npm run build

> my_demo@1.0.0 build
> vite build

vite v5.0.8 building for production...
✓ 22 modules transformed.
[headless-prerender] document is not defined
✓ built in 300ms
error during build:
ReferenceError: document is not defined
    at lt (file:///tmp/headless-prerender-E2BDWq/index-lmWUD0Ml.js:1:129)
    at bt (file:///tmp/headless-prerender-E2BDWq/index-lmWUD0Ml.js:1:17309)
    at jt (file:///tmp/headless-prerender-E2BDWq/index-lmWUD0Ml.js:1:26909)
    at Object.generateBundle (file:///home/arch/ent/my_demo/node_modules/@preact/preset-vite/dist/esm/prerender.mjs:181:38)
    at async Bundle.generate (file:///home/arch/ent/my_demo/node_modules/rollup/dist/es/shared/node-entry.js:16988:9)
    at async file:///home/arch/ent/my_demo/node_modules/rollup/dist/es/shared/node-entry.js:19534:27
    at async catchUnfinishedHookActions (file:///home/arch/ent/my_demo/node_modules/rollup/dist/es/shared/node-entry.js:18965:16)
    at async build (file:///home/arch/ent/my_demo/node_modules/vite/dist/node/chunks/dep-uAHLeuC6.js:66766:22)
    at async CAC.<anonymous> (file:///home/arch/ent/my_demo/node_modules/vite/dist/node/cli.js:845:9)
```

Final `package.json`:

```json
{
  "name": "my_demo",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "type": "module",
  "scripts": {
    "build": "vite build"
  },
  "author": "",
  "license": "ISC",
  "dependencies": {
    "@preact/preset-vite": "file:.yalc/@preact/preset-vite",
    "preact": "^10.19.3",
    "preact-iso": "^2.3.2",
    "preact-render-to-string": "^6.3.1"
  },
  "devDependencies": {
    "typescript": "^5.3.3",
    "vite": "^5.0.8"
  }
}
```

**Reply by ghost** [reply]
Previously, I said:
> nothing broke, but nothing was statically rendered either.

This was not the case, using the demo folder.

When this happened, I was using this setup: https://github.com/RickStanley/poc-ssg-preact/tree/sample

With some modifications to accommodate the new plugin. Like change `src/app.tsx` contents to be equal to `src/index.tsx` of the demo project, moving `src/pages/Home.tsx` to `src/pages/Home/index.tsx` and changing `index.html` (from root) to be the same as the demo, only point to `app.tsx` instead to `index.tsx`.

**Reply by rschristian** [reply]
Will take a look later, but off the top of my head, no clue where that `document` usage is coming from. I think `create-preact` uses Vite 4, might be something added in 5? Haven't touched that yet.

**Reply by rschristian** [reply]
Ah yep, they just broke the preload patch in `5.0.5` a few weeks ago. Here's a working demo: https://github.com/rschristian/demo__preact-prerendering-plugin

**Reply by ghost** [reply]
Nice! I tried it and it worked.

Well then, the SSG part is there. Now I'll try to do some tweaking and see if I can get it to work the way I want.

And since it's not released yet, I may have to copy your plugin over. I don't suppose there's an ETA for this?

Thanks!

**Reply by rschristian** [reply]
No ETA yet, no, though I did mark it as "ready" earlier this morning. We'll see if anyone manages to break it or find issues.

Copying it isn't ideal, but it does work as a standalone plugin for what it's worth. You don't need to copy it into the Vite preset necessarily.

**Reply by ghost** [reply]
When adding:
```
                rollupOptions: {
                        external: ['preact', 'preact/hooks',]
                }
```

And executing `npm run build`, yields:

```bash
vite v5.0.8 building for production...
✓ 19 modules transformed.
[headless-prerender] Cannot find package 'preact' imported from /tmp/headless-prerender-UFzqlJ/index-nDD8L_Im.js
✓ built in 255ms
error during build:
Error: Cannot find package 'preact' imported from /tmp/headless-prerender-UFzqlJ/index-nDD8L_Im.js
    at new NodeError (node:internal/errors:405:5)
    at packageResolve (node:internal/modules/esm/resolve:916:9)
    at moduleResolve (node:internal/modules/esm/resolve:973:20)
    at defaultResolve (node:internal/modules/esm/resolve:1193:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:403:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:372:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:249:38)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:76:39)
    at link (node:internal/modules/esm/module_job:75:36)
```

I can imagine why this is happening. But I wonder: should this setting interfere with pre-rendering at all?

**Reply by rschristian** [reply]
I suppose I need to ask why you're setting `preact` to be external to the app?

**Reply by ghost** [reply]
I want to provide Preact through import map. When adding it to exclude, Vite uses `import` in scripts that consume it.

**Reply by rschristian** [reply]
I suppose we can move the tmp dir local, into `node_modules` or something.

**Reply by rschristian** [reply]
Just to circle back on this, the prerendering functionality was recently released in `@preact/preset-vite` and you can find the [instructions here](https://github.com/preactjs/preset-vite#prerendering-configuration) or create a new project with `create-preact` to get started with prerendering.
